src/cargo/sources/registry/mod.rs

```diff
@@ -552,6 +553,19 @@ impl<'cfg> RegistrySource<'cfg> {
         } else {
             Box::new(remote::RemoteRegistry::new(source_id, config, &name)) as Box<_>
         };
+
+        // Before starting to work on the registry, make sure that
+        // `<cargo_home>/registry` is marked as excluded from indexing and
+        // backups. Older versions of Cargo didn't do this, so we do it here
+        // regardless of whether `<cargo_home>` exists.
+        //
+        // This does not use `create_dir_all_excluded_from_backups_atomic` for
+        // the same reason: we want to exclude it even if the directory already
+        // exists.
+        let registry_base = config.registry_base_path();
+        registry_base.create_dir()?;
```

Comment:
Maybe we can `assert_package_cache_locked` before create_dir?

Reply:
Sorry for the misleading. I don't think we need to acquire lock for `GitSource` and `RegistrySource`. AFAIK, Cargo already requires the lock before entering this.

What I meant is something like

```rust
let registry_base = config.registry_base_path();
registry_base.create_dir()?;
let registry_path = config.assert_package_cache_locked(&registry_base);
exclude_from_backups_and_indexing(&registry_path)
```

No need to call `acquire_package_cache_lock` here.

Reply:
I did try that but I found that quite a few tests fail; I guess there are paths where Cargo gets to here without previously taking a lock.  For example `install::list`:

```

---- install::list stdout ----
running `/home/mbp/src/cargo/target/debug/cargo install --list`
running `/home/mbp/src/cargo/target/debug/cargo install bar --vers =0.2.1`
thread 'install::list' panicked at '
test failed running `/home/mbp/src/cargo/target/debug/cargo install bar --vers =0.2.1`
error: process exited with code 101 (expected 0)
--- stdout

--- stderr
thread 'main' panicked at 'package cache lock is not currently held, Cargo forgot to call `acquire_package_cache_lock` before we got to this stack frame', src/cargo/util/config/mod.rs:1605:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/core/src/panicking.rs:142:14
   2: cargo::util::config::Config::assert_package_cache_locked
             at /home/mbp/src/cargo/src/cargo/util/config/mod.rs:1605:9
   3: cargo::sources::registry::RegistrySource::remote
             at /home/mbp/src/cargo/src/cargo/sources/registry/mod.rs:567:29
   4: cargo::core::source::source_id::SourceId::load
             at /home/mbp/src/cargo/src/cargo/core/source/source_id.rs:305:49
   5: cargo::sources::config::SourceConfigMap::load
             at /home/mbp/src/cargo/src/cargo/sources/config.rs:151:23
   6: cargo::ops::cargo_install::InstallablePackage::new
             at /home/mbp/src/cargo/src/cargo/ops/cargo_install.rs:143:34
   7: cargo::ops::cargo_install::install
             at /home/mbp/src/cargo/src/cargo/ops/cargo_install.rs:572:31
   8: cargo::commands::install::exec
             at /home/mbp/src/cargo/src/bin/cargo/commands/install.rs:163:9
   9: cargo::cli::execute_subcommand
             at /home/mbp/src/cargo/src/bin/cargo/cli.rs:369:16
  10: cargo::cli::main
             at /home/mbp/src/cargo/src/bin/cargo/cli.rs:176:5
  11: cargo::main
             at /home/mbp/src/cargo/src/bin/cargo/main.rs:37:13
  12: core::ops::function::FnOnce::call_once
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
', tests/testsuite/install.rs:945:48
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/core/src/panicking.rs:142:14
   2: cargo_test_support::panic_error::pe
             at ./crates/cargo-test-support/src/lib.rs:46:9
   3: cargo_test_support::panic_error
             at ./crates/cargo-test-support/src/lib.rs:38:5
   4: cargo_test_support::Execs::run
             at ./crates/cargo-test-support/src/lib.rs:800:13
   5: testsuite::install::list
             at ./tests/testsuite/install.rs:945:5
   6: testsuite::install::list::{{closure}}
             at ./tests/testsuite/install.rs:937:1
   7: core::ops::function::FnOnce::call_once
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/core/src/ops/function.rs:248:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

My basic thought was that holding a lock while marking the directory is probably a good idea just on the grounds of consistently doing it on all file IO, but maybe not really necessary for correctness.

I can see a few paths and I would welcome advice:

* look at all the failing tests and work out where they should be taking locks
* mark the directory without taking locks
* (as I did in the earlier draft) split the difference by taking the lock just momentarily, relying on it being reentrant if it's already acquired at a higher level
* ... something else?

Let me know? Thanks for the review.

Reply:
I'm trying to work out from the existing code what is the general style for holding the lock. It seems like, for the sake of allowing as many parallel commands as are safe, it's taken at the lowest level that won't either allow races or cause thrashing? That led me to do it right in this code.

Reply:
I did some quick checks. Hmm… there are indeed somewhere should take locks but didn't. Addressing it seems out of scope of this PR. 

I still prefer to put the logic in corresponding method for each sources. What about back to ce9a6ab6483978b3a4f3b7cc3c86ff5df9609628 and move `CACHEDIR.tag` creation of `RegistrySource` to `block_until_ready`? That aligns more to what we discussed in the issue.

> ...mark these directories every time it runs an operation that updates them, if the mark does not already exist.

Reply:
Please take another look?